File: AncientVisualResearchSuite/languages/csharp/HistoricalScene.cs
namespace AncientVisualResearchSuite;

public record CharacterAgent(string Id,string Role,double X,double Y,double Heading,string Activity);
public sealed class HistoricalScene {
    public List<CharacterAgent> Characters { get; } = new();
    public IEnumerable<IReadOnlyList<CharacterAgent>> Simulate(double seconds,double step=.25) {
        if(seconds<0 || step<=0) throw new ArgumentOutOfRangeException();
        for(double t=0;t<=seconds+1e-9;t+=step)
            yield return Characters.Select(c=>{
                var active=(c.Activity=="idle"||c.Activity=="sleep")?0:.5;
                var phase=c.Heading*Math.PI/180+t*.15;
                return c with { X=c.X+Math.Cos(phase)*active*t, Y=c.Y+Math.Sin(phase)*active*t };
            }).ToArray();
    }
}

File: AncientVisualResearchSuite/languages/java/src/main/java/com/amerhwitat/avrs/HistoricalScene.java
package com.amerhwitat.avrs;
import java.util.*;

public final class HistoricalScene {
  public record CharacterAgent(String id,String role,double x,double y,double heading,String activity) {}
  private final List<CharacterAgent> characters;
  public HistoricalScene(List<CharacterAgent> characters){ this.characters=new ArrayList<>(characters); }
  public List<List<CharacterAgent>> simulate(double seconds,double step){
    if(seconds<0||step<=0) throw new IllegalArgumentException("invalid interval");
    List<List<CharacterAgent>> frames=new ArrayList<>();
    for(double t=0;t<=seconds+1e-9;t+=step){
      List<CharacterAgent> frame=new ArrayList<>();
      for(var c:characters){ double active=(c.activity().equals("idle")||c.activity().equals("sleep"))?0:.5; double phase=Math.toRadians(c.heading())+t*.15;
        frame.add(new CharacterAgent(c.id(),c.role(),c.x()+Math.cos(phase)*active*t,c.y()+Math.sin(phase)*active*t,c.heading(),c.activity())); }
      frames.add(frame);
    }
    return frames;
  }
}
